editor/plugins/curve_editor_plugin.cpp

```diff
@@ -557,43 +672,60 @@ void CurveEditor::update_view_transform() {
 	_world_to_view = view_trans * world_trans;
 }
 
-Vector2 CurveEditor::get_tangent_view_pos(int i, TangentIndex tangent) const {
+Vector2 CurveEdit::get_tangent_view_pos(int p_index, TangentIndex p_tangent) const {
 	Vector2 dir;
-	if (tangent == TANGENT_LEFT) {
-		dir = -Vector2(1, _curve_ref->get_point_left_tangent(i));
+	if (p_tangent == TANGENT_LEFT) {
+		dir = -Vector2(1, curve->get_point_left_tangent(p_index));
 	} else {
-		dir = Vector2(1, _curve_ref->get_point_right_tangent(i));
+		dir = Vector2(1, curve->get_point_right_tangent(p_index));
+	}
+
+	Vector2 point_pos = curve->get_point_position(p_index);
+	Vector2 point_view_pos = get_view_pos(point_pos);
+	Vector2 control_view_pos = get_view_pos(point_pos + dir);
+
+	Vector2 distance_from_point = tangent_length * (control_view_pos - point_view_pos).normalized();
+	Vector2 tangent_view_pos = point_view_pos + distance_from_point;
+
+	// Limit the tangent control point inside the rect.
+	float fraction_inside = 1.0;
+	if (distance_from_point.x != 0.0) {
+		fraction_inside = MIN(fraction_inside, ((distance_from_point.x > 0 ? get_rect().size.x : 0) - point_view_pos.x) / distance_from_point.x);
+	}
+	if (distance_from_point.y != 0.0) {
+		fraction_inside = MIN(fraction_inside, ((distance_from_point.y > 0 ? get_rect().size.y : 0) - point_view_pos.y) / distance_from_point.y);
 	}
 
-	Vector2 point_pos = get_view_pos(_curve_ref->get_point_position(i));
-	Vector2 control_pos = get_view_pos(_curve_ref->get_point_position(i) + dir);
+	if (fraction_inside < 1.0 && fraction_inside > 0.5) {
```

Comment:
Yes. This can happen for points outside the boundary (e.g. a point you put at y=1.05 before changing the range to [0, 1]). Otherwise it'd look very silly.